docs/configuration.md
# Configuring Pivot

It is easy to start using Pivot with Druid by pointing it at your Druid cluster: `pivot --druid druid.broker.host:8082`
but to make the most of Pivot you will want to configure it.

Pivot can be configured with a *config* YAML file. While you could write one from scratch it is recommended to let
Pivot give you a head start by using it to generate a config file for you using the default introspection.

Run:

```bash
pivot --druid druid.broker.host:8082 --print-config --with-comments > config.yaml
```

This will cause Pivot to go through its normal startup and introspection routine and then dump the internally generated
config (complete with comments) into the provided file.

You can now run `pivot --config config.yaml` to run Pivot with your config.

The next step is to open the generated config file in your favourite text editor and configure Pivot to your liking.
Below we will go through a typical configuration flow. At any point you can save the config and re-launch Pivot to load
that config in.

## Configuring the Pivot server

**verbose** (boolean)

Indicates that Pivot should run in verbose mode. This will log all the queries done by Pivot.

**port** (number)

The port that Pivot should run on. Default 9090.

**serverHost** (string)

The host that Pivot will bind to. Defaults to all hosts.

**serverRoot** (string)

A custom path to act as the server string. Default: `pivot`

The Pivot UI will be served from `http://pivot-host:$port/` and `http://pivot-host:$port/$serverRoot`

**iframe** ("allow" | "deny")

Specify whether Pivot will be allowed to run in an iFrame.
If set to 'deny' Pivot will set the following headers:

```
X-Frame-Options: "DENY"
Content-Security-Policy: "frame-ancestors 'none'"
```

This is used to prevent [Clickjacking](http://en.wikipedia.org/wiki/clickjacking).
Learn more about it on [MDN](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/X-Frame-Options).

**trustProxy** ("none" | "always")

Should the server trust the `X-Forwarded-*` headers.

**strictTransportSecurity** ("none" | "always")

Specify that Pivot should set the [StrictTransportSecurity](https://developer.mozilla.org/en-US/docs/Web/Security/HTTP_strict_transport_security) header.

Note that Pivot can itself only run an http server.
This option is intended to be used when when Pivot is running behind a HTTPS terminator like AWS ELB.

## Configuring the Clusters

The top level `clusters:` key that holds the clusters that Pivot will connect to.

### General properties

Each cluster has the following properties:

**name** (string)

The name of the cluster (to be referenced later from the data cube)

**type** ('druid' | 'mysql' | 'postgres')

The database type of the cluster

**host** (string)

The host (hostname:port) of the cluster. In the Druid case this must be the broker.

**version** (string)

The explicit version to use for this cluster.
Define this to override the automatic version detection.

**timeout** (number)

The timeout to set on the queries in ms. Default: 40000

**sourceListScan** ('disable' | 'auto')

Should the sources of this cluster be automatically scanned and new sources added as data cubes. Default: 'auto'

### Druid specific

**introspectionStrategy** (string)

The introspection strategy for the Druid external.

**requestDecorator** (string)

The request decorator module filepath to load.

**decoratorOptions** (any)

Options passed to the request decorator module

### Postgres + MySQL specific

**database** (string)

The database to which to connect to.

**user** (string)

The user to connect as. This user needs no permissions other than SELECT.

**password** (string)

The password to use with the provided user.


## Configuring the DataSources

The top level `dataCubes:` key that holds the data cubes that will be loaded into Pivot.
The order of the data cubes in the config will define the ordering seen in the UI.


### Basic data cube properties

Described here are only the properties which you might want to change.

**name** (string)

The name of the data cube as used internally in Pivot and used in the URLs. This should be a URL safe string.
Changing this property for a given data cube will break any URLs that someone might have generated for that data
cube in the past.

**clusterName** (string)

The cluster that the data cube belongs to (or `'native'` if this is a file based data cube)

**title** (string)

The user visible name that will be used to describe this data cube in the UI. It is always safe to change this.

**description** (string)

The description of the data cube shown in the homepage.

**defaultTimezone** (string - timezone)

The default timezone, expressed as an [Olsen Timezone](https://en.wikipedia.org/wiki/Tz_database),
that will be selected when the user first opens this cube. Default `Etc/UTC`.

**defaultDuration** (string - duration)

The time period, expressed as an [ISO 8601 Duration](https://en.wikipedia.org/wiki/ISO_8601#Durations),
that will be shown when the user first opens this cube. Default `P1D` (1 day).

**defaultSortMeasure** (string)

The name of the measure that will be used for default sorting. It is commonly set to the measure that represents the
count of events. Default: the first measure.

**defaultSelectedMeasures** (string[])

The names of the measures that will be selected by default. Default: first four measures.

**defaultPinnedDimensions** (string[])

The names of the dimensions (in order) that will appear *pinned* by default on the right panel. Default: `[]`.


### Attribute Overrides

While Pivot tries to learn as much as it can from your data cube from Druid directly.
It can not (yet) do a perfect job. The `attributeOverrides:` section of the data cube is there for you to fix that.

**name** (string)

The name of the attribute (column) in Druid. This must match the Druid name.

Here are some common scenarios where you should add an attribute override:

#### You have a HyperLogLog metric column but Pivot is not detecting it

If you have a HyperLogLog metric (say: `unique_things`) it is possible that Druid introspection (Druid <= 0.8.3) will
not describe it correctly.
In that case it will be assumed to be a regular numeric column and will ether return incorrect results (or error out)
when included in queries.

You should add:

```yaml
         - name: unique_things
           special: unique
```

To the `attributeOverrides` to tell Pivot that this is indeed a special (hyperUnique) column.

You should also ensure that wherever it is used in the measures it is aggregated with `countDistinct($unique_things)`.

#### You have a numeric dimension

Pivot can not corretly detect numeric dimensions as Druid reports all dimensions to be strings.
When a numeric dimension is incorrectly classified as a string its soring will appear wrong in the UI.
If you have a dimension with numeric values (say: `age`).

You should add:

```yaml
         - name: age
           type: NUMBER
```

To the `attributeOverrides` to tell Pivot that this is numeric.

You can now use `$age` in numeric expressions. For example you could create a dimension with the formula
`$age / 2 + 7`.


### Dimensions

In this section you can define the dimensions that users can *split* and *filter* on in the UI.
The order of the dimension list in the top of the left panel is determined by the order of the dimensions definitions
in this section.

**name** (string)

The name of the dimension.
This does not have to correspond to the attribute name (but the auto generated dimensions do).
This should be a URL safe string.
Changing this property will break any URLs that someone might have generated that include this dimension.

**title** (string)

The title for this dimension in the UI. Can be anything and is safe to change at any time.

**url** (string)

A url associated with the dimension, with optional token '%s' that is replaced by the dimension value to generate
a link specific to each value.

**granularities** (string[5])

For time dimensions you can define a set of exactly 5 granularities that you want to be available for bucketing.

Each granularity must be expressed as a 'floorable' [ISO 8601 duration](https://en.wikipedia.org/wiki/ISO_8601#Durations)
A floorable duration is any duration that is either a single period like `P1D`, `P1W`, `PT1S`, e.t.c. or a multiple period
that divides into the larger period. For example, `P3M` is floorable because 3 divides 12 but `P5M` is not floorable.

By default `granularities` is set to: `['PT1M', 'PT5M', 'PT1H', 'P1D', 'P1W']`

If you mainly care about smaller intervals, you might want to set it to: `['PT1S', 'PT15S', 'PT30S', 'PT1M', 'PT1H']`

Alternatively, if you mainly care about large intervals, you might want to try: `['P1D', 'P1W', 'P1M', 'P3M', 'P1Y']`

**bucketingStrategy** ('defaultBucket' | 'defaultNoBucket')

Specify whether or not the dimension should be bucketed by default. If unspecified defaults to 'defaultBucket' for time and numeric dimensions.

**sortStrategy** ('self' | `someMeasureName`)

Specify a specific sort strategy for this dimension in visualizations. If unspecified defaults to best sort strategy based on the visualization.

**formula** (string - plywood expression)

The formula for this dimension. By default it is `$name` where *name* is the name of the dimension.
You can create derived dimensions by using non-trivial formulas.

Here are some common use cases for derived dimensions:

#### Lookups

If you have a dimension that represents an ID that is a key into some other table. You may have set up a
[Druid Query Time Lookup](http://druid.io/docs/latest/querying/lookups.html) in which case you could

```yaml
      - name: correctValue
        formula: $lookupKey.lookup('my_awesome_lookup')
```

Which would apply the lookup.

You can also apply the `.fallback()` action as ether:

- `$lookupKey.lookup('my_awesome_lookup').fallback($lookupKey)` to keep values that were not found as they are.
- `$lookupKey.lookup('my_awesome_lookup').fallback('missing')` to map missing values to the word 'missing'.

#### Extraction

Imagine you have an attribute `resourceName` which has values:

```json
["druid-0.8.2", "druid-0.8.1", "druid-0.7.0", "index.html"]
```

You could apply, for example, the `.extract` function by creating the following dimension:

```yaml
      - name: resourceVersion
        formula: $resourceName.extract('(\d+\.\d+\.\d+)')
```

Which would have values:

```json
["0.8.2", "0.8.1", "0.7.0", null]
```

#### Boolean

It is often useful to create dimensions that are the result of some boolean expression.
Let's say that you are responsible for all accounts in the United States as well as some specific account you could create a dimension like:

```yaml
      - name: myAccounts
        formula: $country == 'United States' or $accountName.in(['Toyota', 'Honda'])
```

Now my account would represent a custom filter boolean diemension.

#### Custom transformations

If no existing plywood function meets your needs, you could also define your own custom transformation.
The transformation could be any supported [Druid extraction function](http://druid.io/docs/latest/querying/dimensionspecs.html).

For example you could apply any number of javascript functions to a string.

To use that in Pivot define:

```yaml
    options:
      customTransforms:
        stringFun:
          extractionFn:
            type: javascript
            function: function(x) { try { return decodeURIComponent(x).trim().charCodeAt(0) } catch(e) { return null; } }
```

Then in the dimensions simply reference `stringFun` like so:

```yaml
      - name: stringFun
        title: String Fun
        formula: $countryURL.customTransform('stringFun')
```

### Measures

In this section you can define the measures that users can *aggregate* on (*apply*) on in the UI.
The order of the measure list in the bottom of the left panel is determined by the order of the measure definitions
in this section.

**name** (string)

The name of the measure.
This should be a URL safe string.
Changing this property will break any URLs that someone might have generated that include this measure.

**title** (string)

The title for this measure in the UI. Can be anything and is safe to change at any time.

**units** (string)

The units for this measure. To be shown alongside the title.

**formula** (string - plywood expression)

The [Plywood expression](http://plywood.imply.io/expressions) for this dimension. By default it is `$main.sum($name)` where *name* is the name of the measure.

The `$main` part of the measure expressions serves as a placeholder for the data segment.
In Plywood every aggregate is a function that acts on a data segment.

You can create derived measures by using non-trivial expressions. Here are some common use cases for derived dimensions:


#### Dividing to compute ratios

Ratios are generally considered fun.

```yaml
      - name: ecpm
        title: eCPM
        formula: $main.sum($revenue) / $main.sum($impressions) * 1000
```


#### Filtered aggregations

A very powerful tool is to use a filtered aggregate.
If, for example, your revenue in the US is a very important measure you could express it as:

```yaml
      - name: usa_revenue
        title: USA Revenue
        formula: $main.filter($country == 'United States').sum($revenue)
```

It is also common to express a ratio of something filtered vs unfiltered.

```yaml
      - name: errorRate
        formula: $main.filter($statusCode == 500).sum($requests) / $main.sum($requests)
```


#### Custom aggregations

Within the measures you have access to the full power of the [Plywood expressions](http://plywood.imply.io/expressions).
If you ever find yourself needing to go beyond the expressive potential of Plywood you could define your own custom aggregation.
The aggregation could be any supported Druid aggregation.

For example Plywood currently does not support the modulo operator.
While Druid has no native modulo support ether it is possible to modulo a measure by using a [javascript aggregator](http://druid.io/docs/latest/querying/aggregations.html#javascript-aggregator).

To use that in Pivot define:

```yaml
    options:
      customAggregations:
        addedMod1337:
          aggregation:
            type: javascript
            fieldNames: ['added']
            fnAggregate: "function(current, added) { return (current + added) % 1337 }"
            fnCombine: "function(partialA, partialB) { return (partialA + partialB) % 1337 }"
            fnReset: "function() { return 0; }"
```

Then in the measures simply reference `addedMod1337` like so:

```yaml
      - name: addedMod
        title: Added Mod 1337
        formula: $main.customAggregate('addedMod1337')
```

This functionality can be used to access any custom aggregations that might be loaded via extensions.


#### Switching metric columns

If you switch how you ingest you underlying metric and can't (or do not want to) recalculate all of the previous data,
you could use a derived measure to seemly merge these two metrics in the UI.

Let's say you had a metric called `revenue_in_dollars` and for some reason you will now be ingesting it as `revenue_in_cents`.

Furthermore right now your users are using pivot with the measure:

```yaml
      - name: revenue
        title: Revenue
        formula: $main.sum($revenue_in_dollars)
```

If your data had a 'clean break' where all events have ether `revenue_in_dollars` or `revenue_in_cents` with no overlap you could use:

```yaml
      - name: revenue
        title: Revenue
        formula: $main.sum($revenue_in_dollars) + $main.sum($revenue_in_cents) / 100
```

If instead there was a period where you were ingesting both metrics then the above solution would double count that interval.
You can 'splice' these two metrics together at a specific time point.

Logically you should be able leverage the [Filtered aggregations](#filtered-aggregations) to do:

```yaml
      - name: revenue  # DO NOT DO THIS IT WILL NOT WORK WITH DRUID < 0.9.2
        title: Revenue
        formula: >
          $main.filter(__time < '2016-04-04T00:00:00Z').sum($revenue_in_dollars) +
          $main.filter('2016-04-04T00:00:00Z' <= __time).sum($revenue_in_cents) / 100
```

But the above will not work because, as of this writing, [Druid can not filter on time in measures](https://github.com/druid-io/druid/issues/2816).

Instead you can leverage [Custom aggregations](#custom-aggregations) and the `javascript` aggregation to achieve essentially the same thing:

```yaml
    # Add this to the data cube options
    options:
      customAggregations:
        revenueSplice:
          aggregation:
            type: javascript
            fieldNames: ['__time', 'revenue_in_dollars', 'revenue_in_cents']
            fnAggregate: "function(current, time, revD, revC) { return current + (time < 1442080800000 ? revD : (revC / 100)); }"
            fnCombine: "function(partialA, partialB) { return partialA + partialB; }"
            fnReset: "function() { return 0; }"
```

Then in the measure definitions:

```yaml
      - name: revenue
        title: Revenue
        formula: $main.customAggregate('revenueSplice')
```

Note that whichever method you chose you should not change the `name` attribute of your original measure as it will preserve the function of any bookmarks.

## Customization

You can define a `customization` option in the config to configure some aspects of the look and feel of Pivot.

### Visual

Can customize the header background color and logo icon by supplying a color string and SVG string respectively.

```yaml
  customization:
    customLogoSvg: >
      <svg width="300" height="200"
        xmlns="http://www.w3.org/2000/svg"
        xmlns:xlink="http://www.w3.org/1999/xlink">
        <rect width="100%" height="100%" fill="green" />
      </svg>

    headerBackground: '#2D95CA'
```

### External links

Pivot supports defining external view links with access to `dataCube`, `filter`, `splits`, and `timezone` objects at link generation time.
This is done by defining a function body in the configuration file.

For example:

```yaml
  customization:
    externalViews:
      - title: Timezone Info
        linkGenerator: >
          {
            return 'http://www.tickcounter.com/timezone/' + timezone.toString().toLowerCase().replace(/\//g, '-');
          }
```

These custom links will appear in the share menu.

![custom-link](./images/custom-link.png)

By default, external views are opened in a new tab but you can disable this by setting `sameWindow: true`

### Timezones

You can customize the timezones that appear in the header bar dropdown by providing an array of timezone strings.

For example:

```yaml
  customization:
    timezones: ['Pacific/Niue', 'Pacific/Marquesas', 'America/Tijuana']
```


CHANGELOG.md
# Change Log

For updates follow [@implydata](https://twitter.com/implydata) on Twitter.

## 0.10.1

- Added collections
- Added cold start settings flow
- Settings fully working
- Redid maxTime storage and calculation (added Timekeeper)

## 0.9.39

- Fixed numeric dimension extent computation

## 0.9.38

- Fixed side drawer animation

## 0.9.37

- Dump any and all nodejitsu dependencies added to shrinkwrap
- Contains and Regexp search filters

## 0.9.36

- Added ability to define custom extraction functions
- Added ability to define custom druid context

## 0.9.35

- Fixed unit forwarding

## 0.9.34

- Added ability to customize `logoutHref`

## 0.9.33

- Fix CSS for small screens

## 0.9.32

- Add `--server-host` command line var
- Add `--server-root` command line var

## 0.9.31

- Fix `defaultSplits` in `dataCube` 

## 0.9.30

- New plywood + more back compat tests

## 0.9.29

- Added user_email to request tracker

## 0.9.28

- Added basic request tracking ability
- Changed default request log format and made it configurable

## 0.9.27

- Added gear icon links to settings

## 0.9.26

- DataCube `subsetFilter` -> `subsetFormula`
- Settings fixes

## 0.9.25

- Settings progress

## 0.9.24

- Added trustProxy to server options
- Removed 'when-x-forwarded-proto'

## 0.9.23

- Adding strictTransportSecurity to server options
- Changing ServerSettings to follow new defaults pattern

## 0.9.22

- Moved default user

## 0.9.21

- Renamed `DataSource` to `DataCube`
- Changed Dimension and Measure `expression` to `formula`

## 0.9.20

- Fixes for sorting on dimension when there are multiple splits

## 0.9.19

- Fixes for data source transitions

## 0.9.18

- `DataSource#engine` renamed to `DataSource#clusterName`
- Cleaned up integration tests

## 0.9.17

- Added better small screen support
- Fixed dragging polyfill (on touch devices)
- Added file notices

## 0.9.16

- Selected filtered items appear at top of menu

## 0.9.15

- Allow continuous dimension in two split line chart
- User menu everywhere

## 0.9.14

- New home view
- Added descriptions to data sources

## 0.9.13

- New plywood with fix for: https://github.com/implydata/plywood/pull/121

## 0.9.12

- New plywood fixes bugs

## 0.9.11

- Do not shrinkwrap npm modules

## 0.9.10

- Fix sort ordering of granularities in the granularity picker
- Single measure mode will re-sort on click
- Configurable timezones

## 0.9.9

- Auto generated config links to a github tag
- Fix config auto-generator printing out `defaultSelectedMeasures: []` 

## 0.9.8

- Early detection of (invalid) data sources that have a measure and dimension of the same name
- Prevention of making data sources with measures and dimensions of the same name 

## 0.9.7

- Tiny style fix

## 0.9.6

- Fix sourceListScan introspection to introspect unknown sources as they appear

## 0.9.5

- Fixed display of exclusion filters 
- Ensure cluster connection prior to starting introspection

## 0.9.4

- Exclusion filters

## 0.9.3

- Removed tests, docs, resources, e.t.c from npm module

## 0.9.2

- Change sourceListScan back to default to auto
- Typos and misc error message improvements

## 0.9.1

- Complete refactor of settings architecture
- Settings can be updated at runtime
- Pivot can connect to multiple clusters
- Pivot can connect to Postgres and MySQL
- Full support of Druid 0.9.1
- Auth module versioning
- Command line flag changes (see [writeup](docs/pivot-0.9.x-migration.md))
- Druid Request decorator changes (see [writeup](docs/pivot-0.9.x-migration.md))
- Ability to [generate programmatic links](docs/generating-links-into-pivot.md) into Pivot
- Ability to configure `serverHost`
- Auto refresh icon rotation
- Pivot is npm shrinkwrapped
- Better dimension / measure panel scaling 

## 0.8.42

- Fixed bug in Print config and added CLI tests

## 0.8.41

- New overflow style in split bar also
- Ability to define defaultSelectedMeasures
- Misc fixes for time filter rendering

## 0.8.40

- New Plywood adds support for native quantiles
- New filter menu overflow style

## 0.8.39

- Fixed example config

## 0.8.38

- Continuous dimension filtering and splitting

## 0.8.37

- Fixed small bar bug

## 0.8.36

- Misc query improvements from new Plywood 

## 0.8.35

- Two splits with scrolling in bar chart
- Pivot will only query within server root

## 0.8.34

- New table scroller
- About dialog shows version

## 0.8.33

- Fixed error when going from Time Series to Bar Chart

## 0.8.32

- Configurable server root (default: `/pivot`)

## 0.8.31

- Support and auto detection for Theta sketches (for countDistinct() only for now) 

## 0.8.30

- Horizontal bars in Table
- Misc visualization selection fixes

## 0.8.29

- Side panel resizing

## 0.8.28

- Visualization class refactor
- Table highlight bubble now shows full value
- Better word wrapping in highlight bubbles
- New Plywood ensures to send UTC timestamps to Druid
- Ability to define custom granularities for bucketing

## 0.8.27

- New chronoshift fixes bug with flooring `PT12H`
- Fix burger menu overflow

## 0.8.26

- Better time axis for tiny intervals

## 0.8.25

- New Plywood (0.10.14)

## 0.8.24

- Improved performance of raw data modal
- Added all US timezones

## 0.8.23

- Fixed npm publish issue

## 0.8.22

- Timezone support
- Date range picker
- Fixed X-axis labeling

## 0.8.21

- Build system works with node 6
- Ability to set custom page title

## 0.8.20

- Added pre-calculation pattern to visualizations
- Fixed bar chart on negative values 

## 0.8.19

- More robust cross browser dragging
- Faster incremental build system

## 0.8.18

- Better favicon
- Fixed natural bar chart order

## 0.8.17

- Build `gulp` will fail (exit 1) on error

## 0.8.16

- Fixed sort metric when following auto suggestion

## 0.8.15

- Export data to CSV
- New time series bubbles with hover logic
- Raw data modal allows you to see the raw data in the selected segment

src/server/config.ts
/*
 * Copyright 2015-2016 Imply Data, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import * as path from 'path';
import * as nopt from 'nopt';
import { TRACKER, LOGGER } from 'logger-tracker';

import { arraySum, inlineVars } from '../common/utils/general/general';
import { Cluster, DataCube, SupportedType, AppSettings } from '../common/models/index';
import { MANIFESTS } from '../common/manifests/index';
import { appSettingsToYAML } from '../common/utils/yaml-helper/yaml-helper';
import { ServerSettings } from './models/index';
import { loadFileSync, SettingsManager, SettingsStore } from './utils/index';

const AUTH_MODULE_VERSION = 1;
const PACKAGE_FILE = path.join(__dirname, '../../package.json');

function exitWithMessage(message: string): void {
  console.log(message);

  // Hack: load the package file for no reason other than to make some time for console.log to flush
  try { loadFileSync(PACKAGE_FILE, 'json'); } catch (e) { }

  process.exit();
}

function exitWithError(message: string): void {
  console.error(message);
  process.exit(1);
}

function zeroOne(thing: any): number {
  return Number(Boolean(thing));
}

function appSettingsJSHasOnLoad(appSettingsJS: any): boolean {
  if (appSettingsJS.sourceReintrospectOnLoad || appSettingsJS.sourceListRefreshOnLoad) return true;
  if (Array.isArray(appSettingsJS.clusters)) {
    return appSettingsJS.clusters.some((cluster: any) => cluster.sourceReintrospectOnLoad || cluster.sourceListRefreshOnLoad);
  }
  return false;
}

var packageObj: any = null;
try {
  packageObj = loadFileSync(PACKAGE_FILE, 'json');
} catch (e) {
  exitWithError(`Could not read package.json: ${e.message}`);
}
export const VERSION = packageObj.version;

const USAGE = `
Usage: pivot [options]

Possible usage:

  pivot --examples
  pivot --druid your.broker.host:8082

General arguments:

      --help                   Print this help message
      --version                Display the version number
  -v, --verbose                Display the DB queries that are being made

Server arguments:

  -p, --port <port-number>     The port pivot will run on (default: ${ServerSettings.DEFAULT_PORT})
      --server-host <host>     The host on which to listen on (default: all hosts)
      --server-root <root>     A custom server root to listen on (default ${ServerSettings.DEFAULT_SERVER_ROOT})

Data connection options:

  Exactly one data connection option must be provided.

  -c, --config <path>          Use this local configuration (YAML) file
      --examples               Start Pivot with some example data for testing / demo
  -f, --file <path>            Start Pivot on top of this file based data cube (must be JSON, CSV, or TSV)
  -d, --druid <host>           The Druid broker node to connect to
      --postgres <host>        The Postgres cluster to connect to
      --mysql <host>           The MySQL cluster to connect to

      --user <string>          The cluster 'user' (if needed)
      --password <string>      The cluster 'password' (if needed)
      --database <string>      The cluster 'database' (if needed)

Configuration printing utilities:

      --print-config           Prints out the auto generated config
      --with-comments          Adds comments when printing the auto generated config
`;

function parseArgs() {
  return nopt(
    {
      "help": Boolean,
      "version": Boolean,
      "verbose": Boolean,

      "port": Number,
      "server-host": String,
      "server-root": String,

      "examples": Boolean,
      "example": String, // deprecated
      "config": String,
      "auth": String,

      "print-config": Boolean,
      "with-comments": Boolean,

      "file": String,
      "druid": String,
      "postgres": String,
      "mysql": String,

      "user": String,
      "password": String,
      "database": String
    },
    {
      "v": ["--verbose"],
      "p": ["--port"],
      "c": ["--config"],
      "f": ["--file"],
      "d": ["--druid"]
    },
    process.argv
  );
}

var parsedArgs = parseArgs();

if (parsedArgs['help']) {
  exitWithMessage(USAGE);
}

if (parsedArgs['version']) {
  exitWithMessage(VERSION);
}

if (parsedArgs['example']) {
  delete parsedArgs['example'];
  parsedArgs['examples'] = true;
}

const SETTINGS_INPUTS = ['config', 'examples', 'file', 'druid', 'postgres', 'mysql'];

var numSettingsInputs = arraySum(SETTINGS_INPUTS.map((input) => zeroOne(parsedArgs[input])));

if (numSettingsInputs === 0) {
  exitWithMessage(USAGE);
}

if (numSettingsInputs > 1) {
  console.error(`only one of --${SETTINGS_INPUTS.join(', --')} can be given on the command line`);
  if (parsedArgs['druid'] && parsedArgs['config']) {
    console.error(`Looks like you are using --config and --druid in conjunction with each other`);
    console.error(`This usage is no longer supported. If you are migrating from Pivot < 0.9.x`);
    console.error(`Please visit: (https://github.com/implydata/pivot/blob/master/docs/pivot-0.9.x-migration.md)`);
  }
  process.exit(1);
}

export const PRINT_CONFIG = Boolean(parsedArgs['print-config']);
export const START_SERVER = !PRINT_CONFIG;

if (START_SERVER) LOGGER.init();

// Load server settings
var serverSettingsFilePath = parsedArgs['config'];

if (parsedArgs['examples']) {
  serverSettingsFilePath = path.join(__dirname, `../../config-examples.yaml`);
}

var anchorPath: string;
var serverSettingsJS: any;
if (serverSettingsFilePath) {
  anchorPath = path.dirname(serverSettingsFilePath);
  try {
    serverSettingsJS = inlineVars(loadFileSync(serverSettingsFilePath, 'yaml'), process.env);
    LOGGER.log(`Using config ${serverSettingsFilePath}`);
  } catch (e) {
    exitWithError(`Could not load config from '${serverSettingsFilePath}': ${e.message}`);
  }
} else {
  anchorPath = process.cwd();
  serverSettingsJS = {};
}

if (parsedArgs['port']) {
  serverSettingsJS.port = parsedArgs['port'];
}
if (parsedArgs['server-host']) {
  serverSettingsJS.serverHost = parsedArgs['server-host'];
}
if (parsedArgs['server-root']) {
  serverSettingsJS.serverRoot = parsedArgs['server-root'];
}
if (parsedArgs['auth']) {
  serverSettingsJS.auth = parsedArgs['auth'];
}

export const VERBOSE = Boolean(parsedArgs['verbose'] || serverSettingsJS.verbose);
export const SERVER_SETTINGS = ServerSettings.fromJS(serverSettingsJS);

// --- Tracker --------------------------------

if (START_SERVER) {
  var trackingUrl = SERVER_SETTINGS.getTrackingUrl();
  if (trackingUrl) {
    TRACKER.init(VERSION, trackingUrl, SERVER_SETTINGS.getTrackingContext());
  }
}

// --- Auth -------------------------------

var auth = SERVER_SETTINGS.auth;
var authMiddleware: any = null;
if (auth && auth !== 'none') {
  auth = path.resolve(anchorPath, auth);
  LOGGER.log(`Using auth ${auth}`);
  try {
    var authModule = require(auth);
  } catch (e) {
    exitWithError(`error loading auth module: ${e.message}`);
  }

  if (authModule.version !== AUTH_MODULE_VERSION) {
    exitWithError(`incorrect auth module version ${authModule.version} needed ${AUTH_MODULE_VERSION}`);
  }
  if (typeof authModule.auth !== 'function') exitWithError(`Invalid auth module: must export 'auth' function`);
  authMiddleware = authModule.auth({
    logger: LOGGER,
    tracker: TRACKER,
    verbose: VERBOSE,
    version: VERSION,
    serverSettings: SERVER_SETTINGS
  });
}
export const AUTH = authMiddleware;

// --- Sign of Life -------------------------------
if (START_SERVER) {
  LOGGER.log(`Starting Pivot v${VERSION}`);
  TRACKER.track({
    eventType: 'pivot_init',
    metric: 'init',
    value: 1
  });
}

// --- Location -------------------------------

const CLUSTER_TYPES: SupportedType[] = ['druid', 'postgres', 'mysql'];

var settingsStore: SettingsStore = null;

if (serverSettingsFilePath) {
  var settingsLocation = SERVER_SETTINGS.getSettingsLocation();
  if (settingsLocation) {
    switch (settingsLocation.getLocation()) {
      case 'file':
        var settingsFilePath = path.resolve(anchorPath, settingsLocation.uri);
        if (settingsLocation.getReadOnly()) {
          settingsStore = SettingsStore.fromReadOnlyFile(settingsFilePath, settingsLocation.getFormat());
        } else {
          settingsStore = SettingsStore.fromWritableFile(settingsFilePath, settingsLocation.getFormat());
        }
        break;

      case 'mysql':
        throw new Error('todo'); // ToDo: make this not incomplete.
        //settingsStore = SettingsStore.fromStateStore(require('../../../pivot-mysql-state-store/index.js').stateStoreFactory());
        //break;

      case 'postgres':
        throw new Error('todo');

      default:
        exitWithError(`unknown location '${settingsLocation.location}'`);
    }

  } else {
    // Assume that the config holds the settings this used to be the only way to provide settings
    try {
      var appSettingsFromConfig = AppSettings.fromJS(serverSettingsJS, { visualizations: MANIFESTS });
    } catch (e) {
      exitWithError(`Could not read setting from config file: ${e.message}`);
    }
    settingsStore = SettingsStore.fromTransient(appSettingsFromConfig, appSettingsJSHasOnLoad(serverSettingsJS));
  }
} else {
  var initAppSettings = AppSettings.BLANK;

  // If a file is specified add it as a dataCube
  var fileToLoad = parsedArgs['file'];
  if (fileToLoad) {
    var fileName = path.basename(fileToLoad, path.extname(fileToLoad));
    initAppSettings = initAppSettings.addDataCube(new DataCube({
      name: fileName,
      title: fileName,
      clusterName: 'native',
      source: fileToLoad
    }));
  }

  for (var clusterType of CLUSTER_TYPES) {
    var host = parsedArgs[clusterType];
    if (host) {
      initAppSettings = initAppSettings.addCluster(new Cluster({
        name: clusterType,
        type: clusterType,
        host: host,

        user: parsedArgs['user'],
        password: parsedArgs['password'],
        database: parsedArgs['database']
      }));
    }
  }

  settingsStore = SettingsStore.fromTransient(initAppSettings);
}

export const SETTINGS_MANAGER = new SettingsManager(settingsStore, {
  logger: LOGGER,
  verbose: VERBOSE,
  anchorPath,
  initialLoadTimeout: SERVER_SETTINGS.getPageMustLoadTimeout()
});

// --- Printing -------------------------------

if (PRINT_CONFIG) {
  var withComments = Boolean(parsedArgs['with-comments']);

  SETTINGS_MANAGER.getFullSettings({
    timeout: 10000
  }).then(fullSettings => {
    console.log(appSettingsToYAML(fullSettings.appSettings, withComments, {
      header: true,
      version: VERSION,
      verbose: VERBOSE,
      port: SERVER_SETTINGS.getPort()
    }));
  }).catch((e: Error) => {
    exitWithError("There was an error generating a config: " + e.message);
  });
}
